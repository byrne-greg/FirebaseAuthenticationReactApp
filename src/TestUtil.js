import React from "react";
import { MemoryRouter } from "react-router-dom";

const withTestRouter = Component => <MemoryRouter>{Component}</MemoryRouter>;

export { withTestRouter };
