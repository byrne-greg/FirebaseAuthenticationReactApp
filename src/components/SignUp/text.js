export default {
  signUp: "Sign Up",
  username: "User Name",
  emailAddress: "Email Address",
  password: "Password",
  confirmPassword: "Confirm Password"
};
