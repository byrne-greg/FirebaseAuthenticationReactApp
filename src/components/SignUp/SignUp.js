import React, { useState, useContext } from "react";
import { Link, withRouter } from "react-router-dom";
import { FirebaseContext } from "../Firebase";
import * as ROUTES from "../../constants/routes";
import text from "./text";

const SignUpPage = () => (
  <div>
    <h1>{text.signUp}</h1>
    <SignUpForm />
  </div>
);

const UsernameInput = ({ username, handleChange }) => (
  <>
    <label id="username-label">{text.username}</label>
    <input
      aria-labelledby="username-label"
      id="username-input"
      name="username"
      value={username}
      onChange={handleChange}
      type="text"
      placeholder={text.username}
    />
  </>
);

const EmailInput = ({ email, handleChange }) => (
  <>
    <label id="email-label">{text.emailAddress}</label>
    <input
      aria-labelledby="email-label"
      id="email-input"
      name="email"
      value={email}
      onChange={handleChange}
      type="text"
      placeholder={text.emailAddress}
    />
  </>
);

const PasswordInput = ({ password, handleChange }) => (
  <>
    <label id="password-label">{text.password}</label>
    <input
      aria-labelledby="password-label"
      id="password-input"
      name="password"
      value={password}
      onChange={handleChange}
      type="password"
      placeholder={text.password}
    />
  </>
);

const ConfirmPasswordInput = ({ confirmPassword, handleChange }) => (
  <>
    <label id="confirm-password-label">{text.confirmPassword}</label>
    <input
      aria-labelledby="confirm-password-label"
      id="confirm-password-input"
      name="confirm-password"
      value={confirmPassword}
      onChange={handleChange}
      type="password"
      placeholder={text.confirmPassword}
    />
  </>
);

const ErrorAlert = ({ errorText }) => (
  <div id="submit-error">
    <p>{errorText}</p>
  </div>
);

const SubmitButton = ({ isInvalid }) => (
  <button id="submit-button" disabled={isInvalid} type="submit">
    {text.signUp}
  </button>
);

const SignUpFormComponent = ({ history }) => {
  const [username, setUsername] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [error, setError] = useState(null);
  const firebase = useContext(FirebaseContext);

  const isInvalid =
    !email || !username || !password || password !== confirmPassword;

  const onSubmit = async event => {
    try {
      const authUser = await firebase.createUser(email, password);
      // Create a user in your Firebase realtime database
      firebase.user(authUser.user.uid).set({
        username,
        email
      });
      setUsername("");
      setEmail("");
      setPassword("");
      setConfirmPassword("");
      history.push(ROUTES.HOME);
    } catch (error) {
      setError(error);
    }

    event.preventDefault();
  };

  const handleChange = event => {
    const value = event.target.value;
    switch (event.target.name) {
      case "username":
        setUsername(value);
        break;
      case "email":
        setEmail(value);
        break;
      case "password":
        setPassword(value);
        break;
      case "confirm-password":
        setConfirmPassword(value);
        break;
      default:
        return; // do nothing
    }
  };

  return (
    <form onSubmit={onSubmit}>
      <UsernameInput username={username} handleChange={handleChange} />
      <EmailInput email={email} handleChange={handleChange} />
      <PasswordInput password={password} handleChange={handleChange} />
      <ConfirmPasswordInput
        confirmPassword={confirmPassword}
        handleChange={handleChange}
      />
      <SubmitButton isInvalid={isInvalid} />
      {error && <ErrorAlert errorText={error.message} />}
    </form>
  );
};

const SignUpForm = withRouter(SignUpFormComponent);

const SignUpLink = () => (
  <p>
    {`${text.dontHaveAccount} `}
    <Link to={ROUTES.SIGN_UP}>{text.signUp}</Link>
  </p>
);

export default SignUpPage;

export { SignUpForm, SignUpLink };
