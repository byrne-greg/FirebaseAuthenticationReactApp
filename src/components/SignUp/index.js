import SignUp, { SignUpForm, SignUpLink } from "./SignUp";

export default SignUp;
export { SignUpForm, SignUpLink };
