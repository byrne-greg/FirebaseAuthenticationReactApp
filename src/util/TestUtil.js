import React from "react";
import { MemoryRouter } from "react-router-dom";
import { FirebaseContext } from "../components/Firebase";
import { AuthUserContext } from "../components/Session";

const withMockFirebase = (mockFirebase, Component) => (
  <FirebaseContext.Provider value={mockFirebase}>
    {Component}
  </FirebaseContext.Provider>
);

const withMockAuthUserContext = (value, Component) => (
  <AuthUserContext.Provider value={value}>{Component}</AuthUserContext.Provider>
);

const withTestRouter = Component => <MemoryRouter>{Component}</MemoryRouter>;

export { withTestRouter, withMockFirebase, withMockAuthUserContext };
