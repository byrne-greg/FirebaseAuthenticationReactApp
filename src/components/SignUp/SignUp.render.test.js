import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { withTestRouter, withMockFirebase } from "../../TestUtil";
import SignUp, { SignUpForm, SignUpLink } from ".";

afterEach(cleanup);

describe("SignUp render tests", () => {
  it("page renders as expected", () => {
    // GIVEN

    // WHEN
    const page = render(withTestRouter(<SignUp />));

    // THEN
    expect(page.baseElement).toMatchSnapshot();
  });

  it("link renders as expected", () => {
    // GIVEN

    // WHEN
    const link = render(withTestRouter(<SignUpLink />));

    // THEN
    expect(link.baseElement).toMatchSnapshot();
  });

  xit("form renders with error", () => {
    // GIVEN
    const firebase = {
      login: () => Promise.resolve()
    };

    // WHEN
    const { form, getByLabelText, getByText } = render(
      withTestRouter(withMockFirebase(firebase, <SignInForm />))
    );
    const emailInput = getByLabelText("Email Address");
    fireEvent.change(emailInput, { target: { value: "email" } });
    const passwordInput = getByLabelText("Password");
    fireEvent.change(passwordInput, { target: { value: "password" } });
    const submitButton = getByText("Sign In");
    fireEvent.click(submitButton);

    // THEN
    expect(form.baseElement).toMatchSnapshot();
  });
});
