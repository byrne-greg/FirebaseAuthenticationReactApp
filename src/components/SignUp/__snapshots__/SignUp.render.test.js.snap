// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp render tests link renders as expected 1`] = `
<body>
  <div>
    <p>
      Don't have an account? 
      <a
        href="/signup"
      >
        Sign Up
      </a>
    </p>
  </div>
</body>
`;

exports[`SignUp render tests page renders as expected 1`] = `
<body>
  <div>
    <div>
      <h1>
        Sign Up
      </h1>
      <form>
        <label
          id="username-label"
        >
          User Name
        </label>
        <input
          aria-labelledby="username-label"
          name="username"
          placeholder="User Name"
          type="text"
          value=""
        />
        <label
          id="email-label"
        >
          Email Address
        </label>
        <input
          aria-labelledby="email-label"
          name="email"
          placeholder="Email Address"
          type="text"
          value=""
        />
         
        <label
          id="password-label"
        >
          Password
        </label>
        <input
          aria-labelledby="password-label"
          name="passwordOne"
          placeholder="Password"
          type="password"
          value=""
        />
        <label
          id="confirm-password-label"
        >
          Confirm Password
        </label>
        <input
          aria-labelledby="confirm-password-label"
          name="passwordTwo"
          placeholder="Confirm Password"
          type="password"
          value=""
        />
        <button
          disabled=""
          type="submit"
        >
          Sign Up
        </button>
      </form>
    </div>
  </div>
</body>
`;
